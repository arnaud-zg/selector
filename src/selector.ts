import { NArgument } from '../types/argument'

export const createSelector = (...args: NArgument.TArgument[]) => <TData>(
  data: TData
) =>
  args.reduce((acc, key) => {
    if (typeof key === 'string' || typeof key === 'number') {
      return (acc as any)[key]
    }

    if (typeof key === 'function') {
      return key(acc)
    }
  }, data)
