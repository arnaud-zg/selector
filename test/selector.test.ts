import { createSelector } from '../src'

interface IItem {
  name: string
  isActive?: boolean
  price?: number
}

interface IData {
  items: IItem[]
}

describe('createSelector', () => {
  it('should create selector', () => {
    const firstElementSelector = createSelector(0)
    expect(typeof firstElementSelector).toEqual('function')
    expect(firstElementSelector([])).toBeUndefined()
    expect(firstElementSelector<string[]>([])).toBeUndefined()
  })

  it('should get arg with unsupported argument', () => {
    const firstElementSelector = createSelector(true)
    expect(firstElementSelector([])).toEqual([])
  })

  it('should get nothing', () => {
    const firstElementSelector = createSelector(0)
    expect(firstElementSelector([])).toBeUndefined()
    expect(firstElementSelector<string[]>([])).toBeUndefined()
  })

  it('should select first element', () => {
    const firstElementSelector = createSelector(0)
    expect(firstElementSelector(["It's me"])).toMatchSnapshot()
    expect(
      firstElementSelector<string[]>(["It's me"])
    ).toMatchSnapshot()
  })

  it('should select first item price', () => {
    const data: IData = {
      items: [
        {
          name: 'Sofa',
          price: 100,
        },
      ],
    }
    const firstElementPriceSelector = createSelector(
      'items',
      0,
      (item: IItem) => item.price
    )
    expect(firstElementPriceSelector(data)).toMatchSnapshot()
  })

  it('should select first active item', () => {
    const data: IData = {
      items: [
        {
          name: 'Television',
          isActive: false,
        },
        {
          name: 'Sofa',
          isActive: true,
        },
        {
          name: 'Door',
          isActive: true,
        },
      ],
    }
    const firstActiveElementSelector = createSelector(
      'items',
      (items: IItem[]) => items.find((item: IItem) => item.isActive)
    )
    expect(firstActiveElementSelector(data)).toMatchSnapshot()
  })
})
