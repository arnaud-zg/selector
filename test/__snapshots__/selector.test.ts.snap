// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createSelector should select first active item 1`] = `
Object {
  "isActive": true,
  "name": "Sofa",
}
`;

exports[`createSelector should select first item price 1`] = `100`;
